Added implementation for GridRNN

phvu opened the pull request:
As discussed here: https://github.com/tensorflow/tensorflow/issues/1453

The implementation is generic, users can specify the number of dimensions and various configurations for those dimensions (input/output/priority/non-recurrent). The type of the cells along dimensions can also be selected among LSTM, GRU, vanilla RNN.

Come with unittests for basic types: 2LSTM (tied weights, non-recurrent), 2BasicLSTM and 2RNN.

I made a simple test of Grid2LSTM for character-level language modeling: https://github.com/phvu/grid-lstm-tensorflow/tree/master/char-rnn.

phoaivu committed:
first commit to grid rnn

phoaivu committed:
Merge branch 'master' into enhancement/grid-rnn

phoaivu committed:
added the first tests, and package organization

phoaivu committed:
Merge branch 'master' into enhancement/grid-rnn

phoaivu committed:
finished testing 2LSTM and 2BasicLSTM

phoaivu committed:
Merge branch 'master' into enhancement/grid-rnn

phoaivu committed:
fix for RNN/GRU

phoaivu committed:
added edge cases tests

phoaivu committed:
removed logs, added docs

phoaivu committed:
added gridGRU

phoaivu committed:
resolved conflicts after merge with upstream master

tensorflow-jenkins commented:
Can one of the admins verify this patch?

phoaivu committed:
Merge branch 'master' into enhancement/grid-rnn

ebrevdo commented:
Will take a look tomorrow.

ebrevdo commented on tensorflow/contrib/grid_rnn/__init__.py:
```diff
@@ -0,0 +1,12 @@
+from __future__ import print_function
+
+# pylint: disable=unused-import,wildcard-import, line-too-long
+from tensorflow.contrib.grid_rnn.python.ops.grid_rnn_cell import *
```
add newline

phvu replied:
done

ebrevdo commented on tensorflow/contrib/grid_rnn/python/__init__.py:
```diff
@@ -0,0 +1,3 @@
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
```
add newline and probably also the copyright (remember 2016)

phvu replied:
done

ebrevdo commented on tensorflow/contrib/grid_rnn/python/kernel_tests/grid_rnn_test.py:
```diff
@@ -0,0 +1,324 @@
+# Copyright 2015 Google Inc. All Rights Reserved.
```
2016 here and elsewhere

ebrevdo commented on tensorflow/contrib/grid_rnn/python/kernel_tests/grid_rnn_test.py:
```diff
@@ -0,0 +1,324 @@
+                             m.name: np.array([[0.1, 0.2, 0.3, 0.4, 0.5, 0.6, 0.7, 0.8]])})
+        self.assertEqual(res[0].shape, (1, 2))
+        self.assertEqual(res[1].shape, (1, 8))
+        self.assertAllClose(res[0], [[ 0.36617181, 0.36617181]])
```
how are you getting these numbers?  are they just a smoke test?

if so, convert all these tests to just making sure that they're not nan/inf and verifying that the shapes are correct.

also add checks on g.get_shape() and s.get_shape()

ebrevdo commented on tensorflow/contrib/grid_rnn/python/kernel_tests/grid_rnn_test.py:
```diff
@@ -0,0 +1,324 @@
+        self.assertEqual(cell.state_size, 8)
+        g, s = cell(x, m)
+        sess.run([tf.initialize_all_variables()])
+        res = sess.run([g, s], {x.name: np.array([[1., 1., 1.]]),
```
change x.name to just x (can pass the object itself); here and elsewhere.

ebrevdo commented on tensorflow/contrib/grid_rnn/python/kernel_tests/grid_rnn_test.py:
```diff
@@ -0,0 +1,324 @@
+        x = tf.zeros([1, 3])
+        m = tf.zeros([1, 8])
+        cell = tf.contrib.grid_rnn.Grid2BasicLSTMCell(2, input_size=3)
+        self.assertEqual(cell.output_size, 2)
```
we're getting rid of both input_size and output_size; output_size is not needed and input_size can be replaced with using get_shape() on the input inside the cell.  please remove these from tests and update the cell accordingly.

ebrevdo commented on tensorflow/contrib/grid_rnn/python/kernel_tests/grid_rnn_test.py:
```diff
@@ -0,0 +1,324 @@
+
+        # This cell produces no output
+        cell = tf.contrib.grid_rnn.GridRNNCell(num_units=2, num_dims=2, input_dims=0, output_dims=None,
+                                               non_recurrent_dims=0, non_recurrent_fn=tf.nn.relu)
```
as elsewhere, remove input_dims and output_dims

phvu replied:
`input_dims` and `output_dims` are the dimension indices of the gridRNN cell. These are different from `input_size` and `output_size`.

ebrevdo commented on tensorflow/contrib/grid_rnn/python/ops/__init__.py:
```diff
@@ -0,0 +1,3 @@
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
```
nit: add newlines here and elsewhere; maybe also the license header (2016 etc)

phvu replied:
done

ebrevdo commented on tensorflow/contrib/grid_rnn/python/ops/grid_rnn_cell.py:
```diff
@@ -0,0 +1,353 @@
+          new_output[d.idx] = d.non_recurrent_fn(rnn_cell.linear(args=linear_args,
+                                                                 output_size=conf.num_units, bias=True))
+      else:
+          new_output[d.idx] = d.non_recurrent_fn(rnn_cell.linear(args=linear_args,
```
can you use the contrib.layers.fully_connected here instead?  we're trying to get rid of rnn_cell.linear with fire.

ebrevdo commented on tensorflow/contrib/grid_rnn/python/ops/grid_rnn_cell.py:
```diff
@@ -0,0 +1,353 @@
+
+  def __init__(self, num_units, input_size=None, tied=False, non_recurrent_fn=None):
+    super(Grid2GRUCell, self).__init__(num_units=num_units, input_size=input_size, num_dims=2,
+                                        input_dims=0, output_dims=0, priority_dims=0, tied=tied,
```
nit: whitespace.

ebrevdo commented:
Thanks for the hard work!  Some comments.

phoaivu committed:
added copyright and newlines

phoaivu committed:
clean up _propagate, better way to reuse variables, don't depends on input and output of the cell, updated tests

phvu commented:
@ebrevdo Thanks for the comments. I updated the code.

- Good to know about `input_size` and `output_size`, I updated the code to not depend on those.
However as the `LSTMCell` is still accepting `input_size`, I still keep `input_size` as a parameter in the `cell_fn` callback. Should it be also removed?

- The tests are real. I computed the values using [this](https://docs.google.com/spreadsheets/d/18bkheoJbmMUqdRFrviUy_TiooSjvvpDqiti7hm2EASY/edit#gid=0). Since the initializer is fixed in the tests (weights are initialized to 0.5 or 0.2, depending on the tests; bias are initialized to zero), the output are deterministic. But of course I can switch to lightweight asserts if that is preferred.

ebrevdo commented:
Without a unit test that calls tf.nn.rnn or tf.nn.dynamic_rnn, it's not clear that these cells interact correctly with those methods.  Can you add such a test for your classes?

phvu commented:
ok will do.

phoaivu committed:
added tests for GridLSTM cells with tf.nn.rnn()

phvu commented:
I added the tests for `Grid1LSTM`, `Grid2LSTM`, `Grid3LSTM` (with ReLU), trained with `tf.nn.rnn`. Can you take a look?

ebrevdo commented:
Will look tomorrow-thanks!
On Apr 6, 2016 11:36 AM, "Vu Pham" <notifications@github.com> wrote:

> I added the tests for Grid1LSTM, Grid2LSTM, Grid3LSTM (with ReLU),
> trained with tf.nn.rnn. Can you take a look?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/tensorflow/tensorflow/pull/1665#issuecomment-206506910>
>

ebrevdo commented:
Jenkins, test this please?

gunan commented:
Can one of the admins verify this patch?

ebrevdo commented:
Sorry for not reviewing earlier.  Let's retest this and we can get it merged.

ebrevdo commented:
@phvu can you rebase on HEAD and run the tests?

vrv commented:
@ebrevdo: why does he have to rebase to HEAD?  Do you think there will be a conflict in doing so?

Also, only we can trigger tests.

vrv commented:
@tensorflow-jenkins: test this please

phvu commented:
Cool, tests passed. So I don't need to rebase master for now?

ebrevdo commented:
Excellent, we're good to go.  Thanks for the contribution!  If you have any example code using this we can consider adding it elsewhere in the repo (e.g. under models).

phvu commented:
Cool, thanks for all the help.
Unfortunately I can't share what I am working on with this yet. I have an application of this to character-level modeling (https://github.com/phvu/grid-lstm-tensorflow/tree/master/char-rnn) but it isn't entirely my code so I put it in a separated repo.

I think at some point I can try reproducing some experiments in the paper. Will submit PRs by then.

jstaker7 commented:
Hi @phvu,

Thanks very much for your work. I'm a little confused as to how this can be applied generally. For example, in the paper a 3-LSTM network was applied to image patches of the MNIST dataset where each patch is c * m units long (which is equal to the depth dimension). So we have a 3D grid of c * m vectors. But in your implementation of __call__, we are expecting a tensor of size batch * c * m * num_dims. For MNIST, shouldn't it support b * c * m * dim1_size * dim2_size ? This would require input_dims to be something like [1, 2] (where dim 0 is batch); but I don't see any tests covering cases besides input_dim=0. Am I missing something?

phvu commented:
Hi @jstaker7,
There are some nuances here.
First, the `__call__` method expects the state tensor of `batch_size * ( (c + m) * num_dims)`. When c=m=2 then c*m = c+m, so sorry for the bad test case, I should have use other values.

Second, in the MNIST case, (as far as I understand) I suppose you could use Grid3LSTMCell (https://github.com/tensorflow/tensorflow/blob/master/tensorflow/contrib/grid_rnn/python/ops/grid_rnn_cell.py#L242). This cells receives input and gives output in the first dimension (index 0), and uses LSTM in the other 2 dimensions.
In order to replicate Figure 11 in the paper, you will need to construct 4 3-LSTM cells, each cell will handle 1 scan direction of the input image. However the first hidden LSTM layer receives original pixels as input, while other hidden LSTM layers receives the output of the LSTM just below it.
How to do that is up to you. One simple way is to have a loop to scan the image in a given scan direction (say left-to-right top-to-bottom), which will give a sequence of vectors, and then feed that sequence into the corresponding 3-LSTM cell for that scan direction. For bigger images, you might want to use `tf.scan()`

So your assumption that input_dims should be `[1,2]` is not true. The cells should receive input at dimension 0. For me this is the only reasonable interpretation of the paper, which also take 1-LSTM and 2-LSTM cells into account. (unless the authors release their implementation so that we can do a comparison).

jstaker7 commented:
Hi @phvu,

Thank you, this helps quite a lot. Since each dimension will always have its own LSTM, does that mean we will never use an `input_dims` other than 0?

phvu commented:
To replicate the experiments in the paper, yes. But I am not sure about that generally.

When we set dimension i to be input_dims, it simply means for each __call__, the cell expects an input tensor for dimension i. For non-input dimensions, the cell will take the recurrent values (extracted from the state tensor) as input.

So I guess we can be creative and feed inputs into more than one dimension. I used dimension 0 as input and output dimensions just as a convention. You can as well construct your own GridRNNCell with any configuration.

jstaker7 commented:
Perfect, thanks so much for the clarification. Super helpful and makes a lot of sense.

This might also be interesting to look into in the future. Grid LSTMs were mentioned at it looks like there was a recent merge: https://github.com/tensorflow/tensorflow/issues/2560
